### PR TITLE
Reorganize warnings + Issue #207

### PR DESCRIPTION
**Pull request by milosstojanovic**
@VladUreche, what do you think about organizing warnings this way?

### CODE REVIEW

File: components/plugin/src/miniboxing/plugin/metadata/MiniboxMetadataUtils.scala [hunk 1]
```diff
@@ -168,59 +133,29 @@ trait MiniboxMetadataUtils {
...
-                      if (!tpar.hasAnnotation(SpecializedClass))
-                        suboptimalCodeWarning(pos, s"The following code could benefit from miniboxing specialization " +
-                                                   s"if the type parameter ${tpar.name} of ${tpar.owner.tweakedToString} " +
-                                                   s"""would be marked as "@miniboxed ${tpar.name}" (it would be used to """ +
-                                                   s"instantiate miniboxed type parameter ${p.name} of ${p.owner.tweakedToString})",
-                                                   p.isGenericAnnotated)
-                      else
-                        replaceSpecializedWithMiniboxedWarning(tpar)
+                    (new BackwardWarningForStemClass(p, tpar, pos, inLibrary = !common.isCompiledInCurrentBatch(p))).warn()
+                    (new BackwardWarningForInnerClass(p, tpar, pos, inLibrary = !common.isCompiledInCurrentBatch(p))).warn()
```
Comment (VladUreche) [inline components/plugin/src/miniboxing/plugin/metadata/MiniboxMetadataUtils.scala]: I understand the logic, but it's a bit counter-intuitive: you create both warnings, but you're sure only one will trigger. Why not move everything in two possible warnings: `ForwardWarning` and `BackwardWarning` -- and add the logic to decide what to do in those two classes? It's a more complex refactoring, but it also moves all the logic and special-casing there. IMO, the signatures for the classes should be:

```
case class ForwardWarning(pos: Position, mboxedTypeParam: Symbol, nonMboxedType: Type)
case class BackwardWarning(pos: Position, nonMboxedTypeParam: Symbol, mboxedType: Type)
```

And when you write:

```
if (metadata.miniboxedTParamFlag(p))
  BackwardWarning(pos, nonMboxedTypeParam = p, mboxedType = tpe)
```

You immediately realize it's incorrect: actually, `p` is miniboxed and `tpe` is not, so it's a forward warning :)

**Comment by VladUreche** [comment]
Milos, the work you've done seems like a very good first step, and the code looks much cleaner now. But we should bite the bullet and take the refactoring all the way to just `ForwardWarning` and `BackwardWarning` -- this is when we're 100% compliant with the paper story and the code can't get any cleaner.

**Comment by milosstojanovic** [comment]
Did you think something like this?

File: components/plugin/src/miniboxing/plugin/metadata/MiniboxMetadataWarnings.scala [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  self: MiniboxInjectComponent =>
+
+  import global._
+  import definitions._
+
+  case class ForwardWarning(pos: Position, mboxedTypeParam: Symbol, nonMboxedType: Type) {
+		def warnAndGetSpecInfo(mboxedTpars: Map[Symbol, SpecInfo]): (Symbol, SpecInfo) = {
+			(mboxedTypeParam, nonMboxedType) match {
+				case (p, tpe) if ScalaValueClasses.contains(tpe.typeSymbol) =>
+					(p, Miniboxed(PartialSpec.valueClassRepresentation(tpe.typeSymbol)))
```
Comment (VladUreche) [inline components/plugin/src/miniboxing/plugin/metadata/MiniboxMetadataWarnings.scala]: Strange indenting -- was that intended?

File: components/plugin/src/miniboxing/plugin/metadata/MiniboxMetadataWarnings.scala [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        case (p, tpe) =>
+          (new ForwardWarningForNotSpecificEnoughTypeParam(p, tpe, pos)).warn()
+        (p, Boxed)
+			}
+		}
+  }
+
+  case class BackwardWarning(pos: Position, nonMboxedTypeParam: Symbol, mboxedType: Type) {
+		def warnAndGetSpecInfo(mboxedTpars: Map[Symbol, SpecInfo]): (Symbol, SpecInfo) = {
+			(nonMboxedTypeParam, mboxedType) match {
```
Comment (VladUreche) [inline components/plugin/src/miniboxing/plugin/metadata/MiniboxMetadataWarnings.scala]: Same here.

**Comment by VladUreche** [comment]
@milosstojanovic, LGTM!
Now I understand the problem -- if we only have forward/backward warning, we have to duplicate the logic in both files. Honestly, I was hoping there's less overlap, but, it seems that's not the case...

File: components/plugin/src/miniboxing/plugin/metadata/MiniboxMetadataWarnings.scala (hunk 2 above)
Comment (milosstojanovic) [inline components/plugin/src/miniboxing/plugin/metadata/MiniboxMetadataWarnings.scala]: oops, not really :)

**Comment by milosstojanovic** [comment]
I fixed idents.
Regarding duplication, should we keep current implementation or go back without forward/backward?

**Comment by VladUreche** [comment]
@milosstojanovic Thanks for the wonderful work!

Regarding duplication: to me it looked like the logic for choosing the specialization (`SpecInfo`) and for showing warnings are orthogonal -- for example, the `isUselessWarning` was sprinkled everywhere and I believed it only affects the `ForwardWarning`. Identically, the `MbArray` warning could be factored out separately.

I still think with careful design we can separate the warning and specialization logic, but right now I simply don't have the time to try it out. I will merge the pull request and if you want to give it another shot (maybe having an enum hinting the case for the `BackwardWarning`), please file another PR. Thanks!

**Pull request merged**